### PR TITLE
Convert branch table name column to a new collation for mysql/mssql to support case sensitive because branch names are case sensitive

### LINKED ISSUE #25909: Unable to migrate repos with duplicate case-different tags
### Description

1. Create new migration from https://github.com/pytorch/pytorch, does not matter whether mirror or not.

After some time, it will error:

```
2023/07/16 18:19:53 ...ices/task/migrate.go:60:func1() [E] runMigrateTask[6] by DoerID[1] to RepoID[47] for OwnerID[1] failed: SyncRepoBranchesWithRepo: Error 1062 (23000): Duplicate entry '47-update-test-distribute' for key 'UQE_branch_s'
2023/07/16 18:19:53 ...ervices/task/task.go:51:handler() [E] Run task failed: SyncRepoBranchesWithRepo: Error 1062 (23000): Duplicate entry '47-update-test-distribute' for key 'UQE_branch_s'
```

image

Likely regression from https://github.com/go-gitea/gitea/pull/22743

### Gitea Version

main

### Database

```
mysqld  Ver 11.0.2-MariaDB for osx10.18 on x86_64 (Homebrew)
```

### PR DESCRIPTION
**Pull request by lunny**
Git branch names are case sensitive. Since #22743 introduced branch name in database and mysql/mssql default collation are case insensitive, this PR converts the column to a case sensitive collation.

Fixes: https://github.com/go-gitea/gitea/issues/25909

### CODE REVIEW

File: models/git/branch.go [hunk 1]
```diff
@@ -115,6 +118,22 @@ type Branch struct {
...
+// so we need to set the collation to case-sensitive which is utf8mb4_bin for mysql and Latin1_General_CS_AS for mssql
+// the function is supported by xorm
+func (b *Branch) TableCollations() []*schemas.Collation {
+	if setting.Database.Type.IsMySQL() {
+		return []*schemas.Collation{
+			{Name: "utf8mb4_bin", Column: "name"},
+		}
+	} else if setting.Database.Type.IsMSSQL() {
+		return []*schemas.Collation{
+			{Name: "Latin1_General_CS_AS", Column: "name"},
```
Comment (wxiaoguang) [inline models/git/branch.go:132]: IIRC SQLServer doesn't treat this collation as UTF-8 internally.

So it's better to add a test for it to make sure wide UTF-8 chars (eg: emoji) can be inserted and selected correctly.

Comment (silverwind) [reply models/git/branch.go:131]: Also should check MariaDB as I recall they have different collation names than MySQL and some are not present on both.

**Comment by silverwind** [comment]
Reference to https://github.com/go-gitea/gitea/issues/25909 added.

**Comment by silverwind** [comment]
Might be good to add a test of a repo migration when case-differing branch name, e.g. `Foo` and `foo`.

**Comment by wxiaoguang** [comment]
->  Make table column branch.name case sensitive #28633

**Comment by lunny** [comment]
replaced by #28633

**Pull request closed without merging**